pkg/wasm/cache.go

```diff
@@ -131,35 +131,40 @@ func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration
 		// TODO(bianpengyuan): Add sanity check on downloaded file to make sure it is a valid Wasm module.
 
 		key.checksum = dChecksum
-		// Check if the module has already been stored as local file. If so, avoid writing the file again.
-		if modulePath := c.getEntry(key); modulePath != "" {
-			return modulePath, nil
-		}
-
-		// Materialize the Wasm module into a local file. Use checksum as name of the module.
 		f := filepath.Join(c.dir, fmt.Sprintf("%s.wasm", dChecksum))
-		err = ioutil.WriteFile(f, b, 0644)
-		if err != nil {
+
+		if err := c.addEntry(key, b, f); err != nil {
 			return "", err
 		}
 
-		// Add the downloaded module into the cache map.
-		c.addEntry(key, f)
-
 		return f, nil
 	default:
 		return "", fmt.Errorf("unsupported Wasm module downloading URL scheme: %v", url.Scheme)
 	}
 }
 
-func (c *LocalFileCache) addEntry(key cacheKey, modulePath string) {
+func (c *LocalFileCache) addEntry(key cacheKey, wasmModule []byte, f string) error {
 	c.mux.Lock()
 	defer c.mux.Unlock()
+
+	// Check if the module has already been added. If so, avoid writing the file again.
+	if ce, ok := c.modules[key]; ok {
+		// Update last touched time.
+		ce.last = time.Now()
```

Comment:
last touched by what? how do we know when envoy is done with it? do we have a way monitor draining listeners?

Reply:
> last touched by what?

The time is updated whenever istio-agent processes the ECDS with remote load, which happens on xds reconnection or any envoyfilter update.

> how do we know when envoy is done with it?

We don't have a reliable way to know that the filter is removed, so the current behavior heuristically remove it if there is no ecds update that uses the cache entry for a day.

> do we have a way monitor draining listeners?

Why does listener draining related to this? Wasm VM should only need the module file at start up, and its lifecycle is not affected by listener draining, right? As far as I understand, VM will live so long as there is any filter that is using that VM.

Reply:
What I'm concerned is that we might still have active listeners using an old ECDS resource after a day if it happens that ECDS is never changed but listeners are changed. But I think it should be fine as long as the new VMs not started (which would require another ECDS I suppose) since Wasm just needs this file on start-up only.
Thanks for clarifying.